waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java

```diff
@@ -252,62 +261,143 @@ private boolean isWhitelisted(String databasePrefix, String database) {
+              List<TableMeta> mappedTableMeta = new ArrayList<>();
+              for (TableMeta tableMeta : tables) {
+                if (isWhitelisted(mapping.getDatabasePrefix(), tableMeta.getDbName())) {
+                  mappedTableMeta.add(mapping.transformOutboundTableMeta(tableMeta));
+                }
               }
+              return mappedTableMeta;
+            } catch (TException e) {
+              LOG.warn(errorMessage, e);
+              return null;
             }
-          } catch (TException e) {
-            LOG.warn("Got exception fetching get_table_meta: {}", e.getMessage());
+          };
+          futures.add(executorService.submit(callableTask));
+        }
+
+        Iterator<DatabaseMapping> iterator = databaseMappingsForPattern.keySet().iterator();
+        for (Future<List<TableMeta>> future : futures) {
+          try {
+            DatabaseMapping mapping = iterator.next();
+            long timeout = mapping.getTimeout();
+            List<TableMeta> mappedTableMeta = future.get(timeout, TimeUnit.MILLISECONDS);
+            combined.addAll(mappedTableMeta);
+          } catch (InterruptedException | ExecutionException | TimeoutException | NullPointerException e) {
+            LOG.warn(errorMessage, e);
           }
         }
+        shutdownExecutorService(executorService);
         return combined;
       }
 
       @Override
       public List<String> getAllDatabases(String databasePattern) {
         List<String> combined = new ArrayList<>();
-        for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsByDbPattern(databasePattern)
-            .entrySet()) {
-          try {
-            DatabaseMapping mapping = mappingWithPattern.getKey();
-            String pattern = mappingWithPattern.getValue();
-            List<String> databases = mapping.getClient().get_databases(pattern);
-            for (String database : databases) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
-              }
+        Map<DatabaseMapping, String> databaseMappingsForPattern = databaseMappingsByDbPattern(databasePattern);
+
+        ExecutorService executorService = Executors.newFixedThreadPool(databaseMappingsForPattern.size());
+        List<Future<List<String>>> futures = new ArrayList<>();
+        String errorMessage = "Can't fetch databases by pattern: ";
+
+        for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsForPattern.entrySet()) {
+          Callable<List<String>> callableTask = () -> {
+            try {
+              DatabaseMapping mapping = mappingWithPattern.getKey();
+              String pattern = mappingWithPattern.getValue();
+              List<String> databases = mapping.getClient().get_databases(pattern);
+              return getMappedWhitelistedDatabases(databases, mapping);
+            } catch (TException e) {
```

Comment:
do we need to catch these? I think we can just throw it and they will be in the ExecutionException